staging/src/k8s.io/apimachinery/pkg/runtime/extension.go

```diff
@@ -18,10 +18,88 @@ package runtime
 
 import (
 	"bytes"
-	"encoding/json"
 	"errors"
+	"fmt"
+	"io"
+	"reflect"
+
+	cbordirect "k8s.io/apimachinery/pkg/runtime/serializer/cbor/direct"
+	"k8s.io/apimachinery/pkg/util/json"
+	"sigs.k8s.io/structured-merge-diff/v4/value"
+
+	cbor "github.com/fxamacker/cbor/v2"
 )
 
+// RawExtension intentionally avoids implementing value.UnstructuredConverter for now because the
+// signature of ToUnstructured does not allow returning an error value in cases where the conversion
+// is not possible (content type is unrecognized or bytes don't match content type).
+func (re RawExtension) toUnstructured(rawContentType string) (interface{}, error) {
+	// This replicates the behavior of the original MarshalJSON implementation, which ignores
+	// Object when Raw is non-nil (even if both are non-nil).
+	if re.Raw != nil {
+		switch rawContentType {
+		default:
+			return nil, fmt.Errorf("cannot convert RawExtension with unrecognized content type to unstructured")
+		case ContentTypeJSON:
+			var u interface{}
+			if err := json.Unmarshal(re.Raw, &u); err != nil {
+				return nil, fmt.Errorf("failed to parse RawExtension bytes as JSON: %w", err)
+			}
+			return u, nil
+		case ContentTypeCBOR:
+			var u interface{}
+			if err := cbordirect.Unmarshal(re.Raw, &u); err != nil {
+				return nil, fmt.Errorf("failed to parse RawExtension bytes as CBOR: %w", err)
+			}
+			return u, nil
+		}
+	}
+
+	if re.Object != nil {
+		rv := reflect.ValueOf(re.Object)
+		u, err := value.TypeReflectEntryOf(rv.Type()).ToUnstructured(rv)
+		if err != nil {
+			return nil, err
+		}
+		return u, nil
+	}
+
+	return nil, nil
+}
+
+// selfDescribedCBOR is the CBOR encoding of the head of the "self-described CBOR" tag.
+var selfDescribedCBOR = []byte{0xd9, 0xd9, 0xf7}
+
+var jsonSpace = []byte{' ', '\t', '\r', '\n'}
+
+func (re RawExtension) guessContentType() string {
+	// TODO: Allow for an explicit content type.
```

Comment:
I'm fine with having a non-exported `guessContentType` that never checks the explicit content type.  The caller would then be expected to check it..